lib/Adapter/TolerantParser/Fixer/MemberEmptyLineFixer.php

```diff
@@ -0,0 +1,167 @@
+            TraitUseClause::class,
+            ClassConstDeclaration::class,
+            PropertyDeclaration::class,
+            MethodDeclaration::class,
+        ]);
+
+        return TextEdits::fromTextEdits($this->textEdits($membersMeta));
+    }
+
+    private function gatherMetadata(Node $node, array $nodeTypes): array
+    {
+        $nodesMeta = [];
+        $nodes = [];
+
+        foreach ($node->getDescendantNodes() as $decendantNode) {
+            if (!in_array(get_class($decendantNode), $nodeTypes)) {
+                continue;
+            }
+
+            $nodes[] = $decendantNode;
+        }
+
+        $previousNode = null;
+        foreach ($nodes as $node) {
+            assert($node instanceof Node);
+            $meta = [
+                self::META_SUCCESSOR => false,
+                self::META_NODE_CLASS => get_class($node),
+                self::META_FIRST => false,
+                self::META_PRECEDING_BLANK_LINES => $this->blankLines($node->getLeadingCommentAndWhitespaceText()),
+                self::META_PRECEDING_BLANK_START => $node->getFullStart(),
+                self::META_PRECEDING_BLANK_LENGTH => $node->getStart() - $node->getFullStart(),
+                self::META_INDENTATION => $this->indentation($node),
+                self::META_IS_METHOD => $node instanceof MethodDeclaration,
+            ];
+
+            if (null === $previousNode) {
+                $meta[self::META_FIRST] = true;
+            }
+
+            if ($previousNode && $previousNode === $meta[self::META_NODE_CLASS]) {
+                $meta[self::META_SUCCESSOR] = true;
+            }
+
+            $nodesMeta[] = $meta;
+            $previousNode = $meta[self::META_NODE_CLASS];
+        }
+
+        return $nodesMeta;
+    }
+
+    private function textEdits(array $membersMeta): array
+    {
+        $edits = [];
+
+        foreach ($membersMeta as $meta) {
+            if ($meta[self::META_FIRST] && $meta[self::META_PRECEDING_BLANK_LINES] > 2) {
+                $edits = $this->removeBlankLines($edits, $meta);
+                continue;
+            }
+
+            if (
+                !$meta[self::META_IS_METHOD] &&
+                (
+                    $meta[self::META_SUCCESSOR] &&
+                    $meta[self::META_PRECEDING_BLANK_LINES] > 2
+                )
+            ) {
+                $edits = $this->removeBlankLines($edits, $meta);
+                continue;
+            }
+
+            if (
+                $meta[self::META_FIRST] === false  &&
+                $meta[self::META_SUCCESSOR] === false &&
+                $meta[self::META_PRECEDING_BLANK_LINES] == 2
+            ) {
+                $edits = $this->addBlankLine($edits, $meta);
+                continue;
+            }
+        }
+
+        return $edits;
+    }
+
+    private function blankLines(string $string): int
+    {
+        // TODO: Extract this to a tested utility (perhaps stick it in
+        //       text-document package)
```

Comment:
yes please

Reply:
Rename it as `countBlankLines` and make it return the number of blank lines instead of the number of carriage return ?

Reply:
:+1: it already returns the number of blank lines